### PR TITLE
fix(asc): bound honored Retry-After by the retry cap

### PR DESCRIPTION
**Pull request by rudrankriyam**
> **Stacked on #2076** (`fix/retry-mutations-on-429`), because it edits the same `withRetry` region. Review the last commit only; it retargets to `main` automatically when #2076 merges.

## Problem

The retry engine clamps every computed backoff to `MaxDelay`, but a `Retry-After` from Apple bypasses that clamp entirely and is slept verbatim. There is no upper bound on what a server can ask for.

When App Store Connect answers a rate limit with a long hint, the command simply stops. It sleeps inside the retry loop until the request context expires, and what an operator sees is:

```
$ asc apps list
  (no output for 30 seconds)
Error: retry cancelled: context deadline exceeded
```

The one fact that would explain everything — that Apple rate limited the account and asked to wait an hour — never reaches the operator. It reads like a network hang or a broken timeout, so the natural next step is to retry the command, which produces another silent stall. In CI, with the longer upload budget, the same stall can burn five minutes of runner time per invocation before failing with a message that says nothing about rate limiting.

## Behavior change

`Retry-After` is honored only within the resolved `MaxDelay`:

- **At or under the cap:** waited out exactly, as today. A 5-second hint still waits 5 seconds even though the exponential backoff would have been shorter.
- **Over the cap:** fail immediately, naming both numbers and the knob:

```
Error: rate limited: App Store Connect asked to wait 1h0m0s, exceeding the 30s retry cap (raise ASC_MAX_DELAY to wait longer): rate limited by App Store Connect (status 429): Too many requests (retry after 1h0m0s)
```

The original error is wrapped, so the 429 still reaches exit-code mapping and `errors.As`/`errors.Is` callers.

The alternative — clamp to the cap and retry anyway — was rejected: the server just told us how long it will keep rejecting, so a retry at the cap collects the same 429 and burns the whole retry budget before failing. Reporting the number the operator needs, immediately, is the useful outcome. `ASC_MAX_DELAY` is the documented way to say "I am willing to wait that long."

### Contrast with the itunes deadline guard

#2122 adds a related guard for `internal/itunes`: there, an over-long `Retry-After` is clamped to `ASC_MAX_DELAY` and only fails fast when the hint outlasts the caller's context deadline. That is right for that caller — `asc reviews ratings --all` treats a per-country deadline as a whole-run abort, so preserving partial results matters more than surfacing the exact wait.

The shared engine has no such partial-result contract, and its callers are single operations whose only outcome is success or a message. So this takes the stricter line: the cap is a statement about how long this run will wait, and a request to exceed it is reported rather than approximated. The two guards compose — neither can sleep past its bound.

## Tests

New `internal/asc/client_core_retry_after_test.go`:

- A 60ms `Retry-After` under a 5s cap is honored exactly (elapsed ≥ 60ms, proving it wins over the 1ms base delay).
- A `Retry-After` exactly at the cap is still honored.
- A 1h `Retry-After` against a 30s cap returns after one attempt with no sleep, names both durations, says "rate limited", and still exposes `*APIError` with status 429.
- End-to-end over httptest: a 429 carrying `Retry-After: 3600` against `ASC_MAX_DELAY=2s` makes exactly one request and fails immediately.

The two fail-fast tests run under a 3-second context so a regression fails the suite instead of hanging it — which is exactly what happened when these tests were written against the current code: the run sat for the full hour until it was killed.

One test from #2076 was updated: `TestClientDo_RetriesRateLimitedMutation` served `Retry-After: 1` under a 1ms cap, which is now a fail-fast case. Its cap is raised to 5s so it keeps testing what it meant to — that `Retry-After` wins over the shorter computed backoff.

Commands run: `make build`, `make format`, `make check-docs`, `make lint`, `ASC_BYPASS_KEYCHAIN=1 make test` — all pass. No live API calls.

## Compatibility

- No change for responses without `Retry-After` (the common case) or with a hint inside the cap.
- `ASC_MAX_DELAY` keeps its meaning and now genuinely bounds every wait the engine performs.
- Applies to reads and, via #2076, to rate-limited writes.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved handling of server-provided retry delays, including numeric and date-based values.
  * Honors delays within configured limits while failing promptly when delays exceed retry or context limits.
  * Provides clearer diagnostics for rate limits, retry caps, context deadlines, and upload-server failures.
  * Prevents unnecessary retries, polling, or waiting when a delay cannot be honored.

* **Documentation**
  * Clarified behavior for retry delays exceeding configured or context limits.

* **Tests**
  * Added coverage for parsing, boundary values, uploads, mutations, rate limits, polling, and resilient reads.

### CODE REVIEW

File: internal/asc/client_core.go [hunk 1]
```diff
@@ -396,6 +397,15 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
 			if delay < 0 {
 				delay = expDelay / 2 // minimum delay
 			}
+		case delay > opts.MaxDelay:
+			// The server asked for longer than this run is willing to wait.
+			// Retrying at the cap would just collect the same rejection, and
+			// sleeping the full hint hides the rate limit behind an eventual
+			// context deadline, so report both numbers now.
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go:485]: Fixed in 9ab6fd43. An explicit server delay is now compared with the remaining context deadline before sleeping; if it cannot fit, the client fails immediately while preserving the wrapped API error. The regression proves one request, sub-deadline return, and retained 429 classification.

File: internal/asc/client_core.go [hunk 2]
```diff
@@ -396,6 +397,15 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
...
+			return zero, fmt.Errorf(
+				"rate limited: App Store Connect asked to wait %s, exceeding the %s retry cap (raise ASC_MAX_DELAY to wait longer): %w",
+				delay, opts.MaxDelay, err,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go]: Fixed in 9ab6fd43. Retry-cap errors use rate-limit wording only for HTTP 429; other retryable statuses use status-neutral wording and retain their wrapped service classification. Added a 503 over-cap regression.

**Comment by rudrankriyam** [comment]
@codex review

File: internal/asc/client_core.go [hunk 3]
```diff
@@ -396,6 +398,31 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
...
 			}
+		case delay > opts.MaxDelay:
+			// The server asked for longer than this run is willing to wait.
+			// Retrying at the cap would just collect the same rejection, and
+			// sleeping the full hint hides the response behind an eventual
+			// context deadline, so report both numbers now.
+			category := retryDelayCategory(err)
+			return zero, fmt.Errorf(
+				"%s: App Store Connect asked to wait %s, exceeding the %s retry cap (raise ASC_MAX_DELAY to wait longer): %w",
+				category, delay, opts.MaxDelay, err,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go]: Reproduced on a56a383b: with a 1-hour Retry-After, a 30-second MaxDelay, and a 3-second context, the old ordering returned only the cap advice and omitted the deadline constraint. The fix moves the context-budget check before the cap branch and reports both constraints/remedies when both bind. It also preserves the API error and wraps context.DeadlineExceeded for errors.Is callers. Regression coverage is in TestWithRetry_FailsFastWhenRetryAfterExceedsCap and the context-budget test; focused asc tests, full asc tests, race asc tests, lint, docs, build, and full repository tests pass.

File: internal/asc/client_core.go [hunk 4]
```diff
@@ -418,8 +418,30 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
 		}
 
 		// Calculate delay
-		delay := GetRetryAfter(err)
-		if delay == 0 {
+		retryAfter := GetRetryAfter(err)
+		delay := retryAfter
+		if retryAfter > 0 {
+			if ctxErr := ctx.Err(); ctxErr != nil {
+				return zero, fmt.Errorf("retry cancelled: %w", ctxErr)
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go]: Fixed on current head db1ee7e7. When context cancellation is observed before a positive Retry-After wait, retryCancelledError now unwraps both context and the original retryable error. A regression preserves retryable status, Retry-After, and HTTP 429 classification.

File: internal/asc/client_core.go [hunk 5]
```diff
@@ -396,6 +418,16 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
...
 			}
+		case delay > opts.MaxDelay:
+			// The server asked for longer than this run is willing to wait.
+			// Retrying at the cap would just collect the same rejection, and
+			// sleeping the full hint hides the response behind an eventual
+			// context deadline, so report both numbers now.
+			category := retryDelayCategory(err)
+			return zero, fmt.Errorf(
+				"%s: App Store Connect asked to wait %s, exceeding the %s retry cap (raise ASC_MAX_DELAY to wait longer): %w",
+				category, delay, opts.MaxDelay, err,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go]: Fixed on current head db1ee7e7. The shared retry message now uses upstream-server wording, and presigned upload retry errors use upload-server attribution instead of App Store Connect. The upload regression asserts no ASC attribution and passes.

File: internal/asc/client_core.go [hunk 6]
```diff
@@ -434,6 +456,16 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
 			if delay < 0 {
 				delay = expDelay / 2 // minimum delay
 			}
+		case delay > opts.MaxDelay:
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go:481]: Fixed on current head db1ee7e7. RetryOptions and docs/API_NOTES now state that Retry-After is honored only through ASC_MAX_DELAY; over-cap hints fail fast with the requested delay, cap, and raise-ASC_MAX_DELAY guidance, including presigned uploads.

File: internal/asc/client_core.go [hunk 7]
```diff
@@ -380,8 +380,30 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
...
+			if ctxErr := ctx.Err(); ctxErr != nil {
+				return zero, fmt.Errorf("retry cancelled: %w", ctxErr)
+			}
+			if deadline, ok := ctx.Deadline(); ok {
+				remaining := time.Until(deadline)
+				if delay >= remaining {
+					if delay > opts.MaxDelay {
+						return zero, fmt.Errorf(
+							"%s: App Store Connect asked to wait %s, exceeding the %s retry cap and the context deadline (%s remaining); raise ASC_MAX_DELAY and the request timeout to wait longer: %w: %w",
+							retryDelayCategory(err), delay, opts.MaxDelay, remaining.Round(time.Millisecond), err, context.DeadlineExceeded,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go]: Verified fixed on current head db1ee7e7. Pre-expiry cap and context-budget failures retain the deadline diagnostic without wrapping context.DeadlineExceeded, so shared RetryReadWithFreshTimeout does not reissue the request. The shared bounded-retry regression passes.

File: internal/asc/client_http.go [hunk 8]
```diff
@@ -453,8 +453,20 @@ func parseRetryAfterHeader(value string) time.Duration {
...
 
 	// Try to parse as seconds first
-	if seconds, err := strconv.Atoi(value); err == nil && seconds > 0 {
+	const maxRetryAfterDuration = time.Duration(1<<63 - 1)
+	if seconds, err := strconv.ParseInt(value, 10, 64); err == nil && seconds > 0 {
+		if seconds > int64(maxRetryAfterDuration/time.Second) {
+			return maxRetryAfterDuration
+		}
 		return time.Duration(seconds) * time.Second
+	} else if _, err := strconv.ParseUint(value, 10, 64); err != nil {
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_http.go]: Verified fixed on current head db1ee7e7. Positive numeric Retry-After values that overflow signed parsing but fit uint64 now saturate to the maximum duration, including a leading plus sign, so the retry cap rejects them. Boundary regressions pass.

File: internal/asc/client_core.go [hunk 9]
```diff
@@ -418,8 +418,30 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
...
+							retryDelayCategory(err), delay, opts.MaxDelay, remaining.Round(time.Millisecond), err,
+						)
+					}
+					return zero, fmt.Errorf(
+						"%s: App Store Connect asked to wait %s, which cannot be honored before the context deadline (%s remaining): %w",
+						retryDelayCategory(err), delay, remaining.Round(time.Millisecond), err,
+					)
+				}
+			}
+		}
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go:463]: Fixed on current head db1ee7e7. When context cancellation is observed before a positive Retry-After wait, retryCancelledError now unwraps both context and the original retryable error. A regression preserves retryable status, Retry-After, and HTTP 429 classification.

File: internal/asc/client_http.go [hunk 10]
```diff
@@ -453,8 +453,20 @@ func parseRetryAfterHeader(value string) time.Duration {
...
+		var numberErr *strconv.NumError
+		if errors.As(err, &numberErr) && errors.Is(numberErr.Err, strconv.ErrRange) && !strings.HasPrefix(value, "-") {
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_http.go]: Verified fixed on current head db1ee7e7. Retry-After saturation now requires a positive all-decimal value, including signed and uint64 overflow cases; malformed trailing text remains zero and falls back to exponential backoff. Boundary regression coverage passes.

File: internal/asc/client_core.go [hunk 11]
```diff
@@ -434,6 +458,16 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
...
 			}
+		case delay > opts.MaxDelay:
+			// The server asked for longer than this run is willing to wait.
+			// Retrying at the cap would just collect the same rejection, and
+			// sleeping the full hint hides the response behind an eventual
+			// context deadline, so report both numbers now.
+			category := retryDelayCategory(err)
+			return zero, fmt.Errorf(
+				"%s: upstream server asked to wait %s, exceeding the %s retry cap (raise ASC_MAX_DELAY to wait longer): %w",
+				category, delay, opts.MaxDelay, err,
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go]: Fixed in 131f4db49. Over-cap and context-budget Retry-After diagnostics now carry a typed terminal marker; IsTransientWaitError stops replaying them while IsRetryable, GetRetryAfter, and typed status remain available to direct callers. Added a PollUntilTolerant regression proving one check and no replay.

File: docs/API_NOTES.md [hunk 12]
```diff
@@ -93,8 +93,8 @@ Finance reports use Apple fiscal months (`YYYY-MM`), not calendar months.
...
+- Retry-After headers are honored when they do not exceed `ASC_MAX_DELAY`; an explicit hint above that cap fails fast with the requested delay, the cap, and guidance to raise `ASC_MAX_DELAY` rather than sleeping at the cap. Configure retry settings via `ASC_MAX_RETRIES`, `ASC_BASE_DELAY`, `ASC_MAX_DELAY`, `ASC_RETRY_LOG`.
+- Uploads to the presigned URLs Apple returns in `uploadOperations` retry per part rather than per file: a PUT part is retried on 408/429/500/502/503/504 and on transient transport failures, using the same retry settings and honoring Retry-After only up to `ASC_MAX_DELAY`. Over-cap hints fail fast; parts that use any other method are never replayed, and each attempt is bounded by `ASC_UPLOAD_TIMEOUT`. This applies to build, screenshot, Game Center, App Clip, subscription, in-app purchase, and app event asset uploads.
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply docs/API_NOTES.md]: Fixed in 131f4db49. API_NOTES now documents that Retry-After hints must fit both ASC_MAX_DELAY and the remaining request context; hints beyond either limit fail fast.

File: internal/asc/poll.go [hunk 13]
```diff
@@ -138,6 +138,10 @@ func IsTransientWaitError(ctx context.Context, err error) bool {
 	if ctx != nil && ctx.Err() != nil {
 		return false
 	}
+	var delayErr *retryDelayExceededError
+	if errors.As(err, &delayErr) {
+		return false
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/poll.go]: Fixed in ecfb0443b. VerifyBuildUploadAfterCommit now treats IsRetryDelayExceeded as a best-effort transient lookup outcome, preserving asynchronous-success behavior through the verification window; regression covers a Retry-After that exceeds the remaining verification context.

File: internal/asc/client_core.go [hunk 14]
```diff
@@ -418,8 +436,30 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
 		}
 
 		// Calculate delay
-		delay := GetRetryAfter(err)
-		if delay == 0 {
+		retryAfter := GetRetryAfter(err)
+		delay := retryAfter
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go:443]: Fixed in ecfb0443b. Retry-After classification now runs before the final retry-budget check; a final over-cap hint receives the terminal marker and remains terminal to PollUntilTolerant, while ordinary final attempts retain the budget-exhaustion marker. Added a regression for 503 then final-attempt 429.

File: internal/asc/client_core.go [hunk 15]
```diff
@@ -434,6 +478,24 @@ func withRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOptio
...
+		case delay > opts.MaxDelay:
+			// The server asked for longer than this run is willing to wait.
+			// Retrying at the cap would just collect the same rejection, and
+			// sleeping the full hint hides the response behind an eventual
+			// context deadline, so report both numbers now.
+			category := retryDelayCategory(err)
+			return zero, &retryDelayExceededError{err: fmt.Errorf(
+				"%s: upstream server asked to wait %s, exceeding the %s retry cap (raise ASC_MAX_DELAY to wait longer): %w",
+				category, delay, opts.MaxDelay, err,
+			)}
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/asc/client_core.go:490]: Fixed in 941b1bd6b and covered end to end in b786b90b3. IsTransientMutationError rejects IsRetryDelayExceeded before IsRetryable, and the equalize integration regression proves a 1h Retry-After results in one POST, one negative readback, no replay, and no long recovery sleep.

File: internal/cli/shared/build_wait_test.go [hunk 16]
```diff
@@ -629,6 +629,41 @@ func TestVerifyBuildUploadAfterCommitIgnoresRetryableLookupErrorsUntilBuildLinks
...
+		}
+		lookupCalls++
+		response, err := buildWaitJSONStatusResponse(http.StatusTooManyRequests, `{
+			"errors": [{"status": "429", "code": "RATE_LIMIT_EXCEEDED", "title": "Too many requests"}]
+		}`)
+		if err != nil {
+			return nil, err
+		}
+		response.Header.Set("Retry-After", "1")
+		return response, nil
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/shared/build_wait_test.go]: Fixed in b786b90b3. The regression now uses httptest.NewServer (with a narrow URL-rewrite transport only because the client intentionally pins ASC BaseURL), preserving GET/path assertions, the 429 payload, Retry-After header, lookup count, and timing.

File: internal/cli/shared/build_wait_test.go [hunk 17]
```diff
@@ -629,6 +631,57 @@ func TestVerifyBuildUploadAfterCommitIgnoresRetryableLookupErrorsUntilBuildLinks
...
+	start := time.Now()
+	err = VerifyBuildUploadAfterCommit(context.Background(), client, "app-1", "upload-current", time.Millisecond, verifyTimeout)
+	elapsed := time.Since(start)
+	if err != nil {
+		t.Fatalf("VerifyBuildUploadAfterCommit() error: %v", err)
+	}
+	if lookupCalls != 1 {
+		t.Fatalf("expected one best-effort upload lookup before honoring Retry-After, got %d", lookupCalls)
+	}
+	if elapsed >= verifyTimeout/2 {
```
[the remark this answers is not in the document]
Comment (rudrankriyam) [reply internal/cli/shared/build_wait_test.go]: Fixed in 34ddbbd30. Removed the tight wall-clock assertions from both regressions; the verification context remains bounded and request counts assert the terminal behavior without scheduler-sensitive timing.
